File: sdk/go/provider/ockito.go
package provider

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"net/url"
	"strings"

	http "github.com/bogdanfinn/fhttp"
)

const ockitoBase = "https://ockito.com/web-api"

func ockitoAnyString(v any) string {
	switch x := v.(type) {
	case string:
		return strings.TrimSpace(x)
	case fmt.Stringer:
		return strings.TrimSpace(x.String())
	case nil:
		return ""
	default:
		return strings.TrimSpace(fmt.Sprint(v))
	}
}

func ockitoJSON(method, u string, headers map[string]string, body []byte) (int, map[string]any, error) {
	var reader io.Reader
	if body != nil {
		reader = bytes.NewReader(body)
	}
	req, err := http.NewRequest(method, u, reader)
	if err != nil {
		return 0, nil, err
	}
	req.Header.Set("Accept", "application/json")
	if body != nil {
		req.Header.Set("Content-Type", "application/json")
	}
	for k, v := range headers {
		req.Header.Set(k, v)
	}

	resp, err := HTTPClient().Do(req)
	if err != nil {
		return 0, nil, err
	}
	defer resp.Body.Close()

	raw, err := io.ReadAll(resp.Body)
	if err != nil {
		return resp.StatusCode, nil, err
	}
	var out map[string]any
	if len(raw) > 0 {
		if err := json.Unmarshal(raw, &out); err != nil {
			return resp.StatusCode, nil, fmt.Errorf("ockito invalid JSON: %s HTTP %d", u, resp.StatusCode)
		}
	} else {
		out = map[string]any{}
	}
	return resp.StatusCode, out, nil
}

func ockitoEncodeToken(accessToken, refreshToken string) string {
	raw, _ := json.Marshal(map[string]string{
		"access_token":  accessToken,
		"refresh_token": refreshToken,
	})
	return string(raw)
}

func ockitoDecodeToken(token string) (string, string, error) {
	value := strings.TrimSpace(token)
	if value == "" || !strings.HasPrefix(value, "{") {
		return "", "", fmt.Errorf("ockito: invalid session token")
	}
	var payload map[string]any
	if err := json.Unmarshal([]byte(value), &payload); err != nil {
		return "", "", fmt.Errorf("ockito: invalid session token")
	}
	accessToken := ockitoAnyString(payload["access_token"])
	refreshToken := ockitoAnyString(payload["refresh_token"])
	if accessToken == "" || refreshToken == "" {
		return "", "", fmt.Errorf("ockito: invalid session token")
	}
	return accessToken, refreshToken, nil
}

func ockitoRefreshAccessToken(refreshToken string) (string, error) {
	status, data, err := ockitoJSON(http.MethodPost, ockitoBase+"/grefresh", map[string]string{
		"Authorization": "Bearer " + refreshToken,
		"X-PASSTHROUGH": "Y",
	}, nil)
	if err != nil {
		return "", err
	}
	if status < 200 || status >= 300 {
		return "", fmt.Errorf("ockito grefresh http %d", status)
	}
	accessToken := ockitoAnyString(data["access_token"])
	if accessToken == "" {
		return "", fmt.Errorf("ockito: invalid refresh response")
	}
	return accessToken, nil
}

func ockitoFetchBearerJSON(u string, accessTokenState *string, refreshToken string) (map[string]any, error) {
	status, data, err := ockitoJSON(http.MethodGet, u, map[string]string{
		"Authorization": "Bearer " + *accessTokenState,
	}, nil)
	if err != nil {
		return nil, err
	}
	if status == 401 {
		refreshed, err := ockitoRefreshAccessToken(refreshToken)
		if err != nil {
			return nil, err
		}
		*accessTokenState = refreshed
		status, data, err = ockitoJSON(http.MethodGet, u, map[string]string{
			"Authorization": "Bearer " + refreshed,
		}, nil)
		if err != nil {
			return nil, err
		}
	}
	if status < 200 || status >= 300 {
		return nil, fmt.Errorf("ockito http %d", status)
	}
	return data, nil
}

func ockitoFlattenInboxRow(raw map[string]any, recipient string) map[string]any {
	return map[string]any{
		"id":          ockitoAnyString(raw["uid"]),
		"from":        ockitoAnyString(raw["sender"]),
		"to":          recipient,
		"subject":     ockitoAnyString(raw["subject"]),
		"text":        ockitoAnyString(raw["snippet"]),
		"html":        ockitoAnyString(raw["html"]),
		"date":        raw["timestamp"],
		"is_read":     false,
		"attachments": []any{},
	}
}

func ockitoFlattenMessage(raw map[string]any, recipient string) map[string]any {
	obj, _ := raw["obj"].(map[string]any)
	if obj == nil {
		obj = raw
	}
	from := ockitoAnyString(obj["sender_email"])
	if from == "" {
		from = ockitoAnyString(obj["SenderEmail"])
	}
	if from == "" {
		from = ockitoAnyString(obj["from_"])
	}
	if from == "" {
		from = ockitoAnyString(obj["From"])
	}
	if from == "" {
		from = ockitoAnyString(obj["from"])
	}
	if from == "" {
		from = ockitoAnyString(obj["sender_name"])
	}
	if from == "" {
		from = ockitoAnyString(obj["SenderName"])
	}
	to := ockitoAnyString(obj["to"])
	if to == "" {
		to = ockitoAnyString(obj["To"])
	}
	if to == "" {
		to = recipient
	}
	return map[string]any{
		"id":          ockitoAnyString(raw["uid"]),
		"from":        from,
		"to":          to,
		"subject":     ockitoAnyString(obj["subject"]),
		"text":        ockitoAnyString(obj["text"]),
		"html":        ockitoAnyString(obj["html"]),
		"date":        obj["date"],
		"is_read":     false,
		"attachments": []any{},
	}
}

func OckitoGenerate() (*CreatedMailbox, error) {
	status, login, err := ockitoJSON(http.MethodPost, ockitoBase+"/gtoken", nil, []byte("{}"))
	if err != nil {
		return nil, err
	}
	if status < 200 || status >= 300 {
		return nil, fmt.Errorf("ockito gtoken http %d", status)
	}

	accessToken := ockitoAnyString(login["access_token"])
	refreshToken := ockitoAnyString(login["refresh_token"])
	if accessToken == "" || refreshToken == "" {
		return nil, fmt.Errorf("ockito: invalid token response")
	}

	status, emailResp, err := ockitoJSON(http.MethodGet, ockitoBase+"/email", map[string]string{
		"Authorization": "Bearer " + accessToken,
	}, nil)
	if err != nil {
		return nil, err
	}
	if status < 200 || status >= 300 {
		return nil, fmt.Errorf("ockito email http %d", status)
	}

	email := ockitoAnyString(emailResp["email"])
	if email == "" {
		if nested, ok := emailResp["email"].(map[string]any); ok {
			email = ockitoAnyString(nested["email"])
		}
	}
	if email == "" || !strings.Contains(email, "@") {
		return nil, fmt.Errorf("ockito: invalid email response")
	}

	return &CreatedMailbox{
		Channel: "ockito",
		Email:   email,
		Token:   ockitoEncodeToken(accessToken, refreshToken),
	}, nil
}

func OckitoGetEmails(token, email string) ([]NormEmail, error) {
	accessToken, refreshToken, err := ockitoDecodeToken(token)
	if err != nil {
		return nil, err
	}
	address := strings.TrimSpace(email)
	if address == "" {
		return nil, fmt.Errorf("ockito: empty email")
	}

	data, err := ockitoFetchBearerJSON(
		ockitoBase+"/retrieve/"+url.PathEscape(address)+"/emails",
		&accessToken,
		refreshToken,
	)
	if err != nil {
		return nil, err
	}
	rows, _ := data["inbox"].([]any)
	out := make([]NormEmail, 0, len(rows))
	for _, row := range rows {
		raw, ok := row.(map[string]any)
		if !ok {
			continue
		}
		uid := ockitoAnyString(raw["uid"])
		if uid == "" {
			out = append(out, NormalizeMap(ockitoFlattenInboxRow(raw, address), address))
			continue
		}

		detail, err := ockitoFetchBearerJSON(
			ockitoBase+"/retrieve/"+url.PathEscape(address)+"/"+url.PathEscape(uid),
			&accessToken,
			refreshToken,
		)
		if err != nil {
			out = append(out, NormalizeMap(ockitoFlattenInboxRow(raw, address), address))
			continue
		}
		out = append(out, NormalizeMap(ockitoFlattenMessage(detail, address), address))
	}
	return out, nil
}

File: sdk/npm/src/providers/ockito.ts
import { Channel, Email, InternalEmailInfo } from "../types";
import { normalizeEmail } from "../normalize";
import { fetchWithTimeout } from "../retry";

const CHANNEL: Channel = "ockito";
const BASE_URL = "https://ockito.com/web-api";

interface LoginResponse {
  access_token?: string;
  refresh_token?: string;
  user_type?: string;
  token_type?: string;
  provider?: string | null;
}

interface AccessTokenResponse {
  access_token?: string;
  token_type?: string;
}

interface EmailResponse {
  email?: string | { email?: string };
}

interface InboxItem {
  uid?: string | number;
  timestamp?: string | number;
  sender?: string;
  subject?: string;
  snippet?: string;
}

interface InboxResponse {
  email?: { email?: string } | string;
  inbox?: InboxItem[];
}

interface MailMessageObject {
  uid?: string | number;
  sender_email?: string;
  sender_name?: string;
  from_?: string;
  from?: string;
  to?: string;
  subject?: string;
  date?: string | number;
  html?: string;
  text?: string;
  SenderEmail?: string;
  SenderName?: string;
  From?: string;
  To?: string;
  Subject?: string;
  Date?: string | number;
}

interface MailMessageResponse {
  uid?: string | number;
  self?: string;
  obj?: MailMessageObject | null;
}

interface PackedToken {
  access_token: string;
  refresh_token: string;
}

async function fetchJSON<T>(
  url: string,
  init?: RequestInit,
): Promise<{ status: number; json: T }> {
  const response = await fetchWithTimeout(url, {
    ...init,
    headers: {
      Accept: "application/json",
      ...(init?.headers || {}),
    },
  });
  const text = await response.text();
  let json: T;
  try {
    json = text ? (JSON.parse(text) as T) : ({} as T);
  } catch {
    throw new Error(`ockito invalid JSON: ${url} HTTP ${response.status}`);
  }
  return { status: response.status, json };
}

function encodeToken(accessToken: string, refreshToken: string): string {
  return JSON.stringify({
    access_token: accessToken,
    refresh_token: refreshToken,
  });
}

function decodeToken(token: string): PackedToken {
  const value = String(token || "").trim();
  if (!value) {
    throw new Error("ockito: invalid session token");
  }
  try {
    const parsed = JSON.parse(value) as Partial<PackedToken>;
    const accessToken = String(parsed.access_token || "").trim();
    const refreshToken = String(parsed.refresh_token || "").trim();
    if (!accessToken || !refreshToken) {
      throw new Error("ockito: invalid session token");
    }
    return { access_token: accessToken, refresh_token: refreshToken };
  } catch {
    throw new Error("ockito: invalid session token");
  }
}

async function refreshAccessToken(refreshToken: string): Promise<string> {
  const response = await fetchJSON<AccessTokenResponse>(
    `${BASE_URL}/grefresh`,
    {
      method: "POST",
      headers: {
        Authorization: `Bearer ${refreshToken}`,
        "X-PASSTHROUGH": "Y",
      },
    },
  );
  if (response.status < 200 || response.status >= 300) {
    throw new Error(`ockito grefresh http ${response.status}`);
  }
  const accessToken = String(response.json?.access_token || "").trim();
  if (!accessToken) {
    throw new Error("ockito: invalid refresh response");
  }
  return accessToken;
}

async function fetchBearerJSON<T>(
  url: string,
  accessTokenState: { accessToken: string },
  refreshToken: string,
  init?: RequestInit,
): Promise<T> {
  const request = async (
    accessToken: string,
  ): Promise<{ status: number; json: T }> =>
    fetchJSON<T>(url, {
      ...init,
      headers: {
        ...(init?.headers || {}),
        Authorization: `Bearer ${accessToken}`,
      },
    });

  let response = await request(accessTokenState.accessToken);
  if (response.status === 401) {
    accessTokenState.accessToken = await refreshAccessToken(refreshToken);
    response = await request(accessTokenState.accessToken);
  }
  if (response.status < 200 || response.status >= 300) {
    throw new Error(`ockito http ${response.status}`);
  }
  return response.json;
}

function normalizeEmailAddress(value: unknown): string {
  if (typeof value === "string") {
    return value.trim();
  }
  if (value && typeof value === "object") {
    const nested = value as { email?: unknown };
    return typeof nested.email === "string" ? nested.email.trim() : "";
  }
  return "";
}

function flattenInboxRow(
  raw: Record<string, unknown>,
  recipient: string,
): Record<string, unknown> {
  return {
    id: raw.uid || raw.id || "",
    from: raw.sender || raw.sender_email || raw.from || "",
    to: recipient,
    subject: raw.subject || "",
    text: raw.snippet || raw.text || "",
    html: raw.html || "",
    date: raw.timestamp || raw.received_at || raw.date || "",
    attachments: [],
  };
}

function flattenMessage(
  raw: MailMessageResponse | Record<string, unknown>,
  recipient: string,
): Record<string, unknown> {
  const obj =
    (raw as MailMessageResponse)?.obj &&
    typeof (raw as MailMessageResponse).obj === "object"
      ? ((raw as MailMessageResponse).obj as MailMessageObject)
      : (raw as Record<string, unknown>);
  return {
    id: (raw as MailMessageResponse)?.uid || obj.uid || "",
    from:
      obj.sender_email ||
      obj.SenderEmail ||
      obj.from_ ||
      obj.From ||
      obj.from ||
      obj.sender_name ||
      obj.SenderName ||
      "",
    to: obj.to || obj.To || recipient,
    subject: obj.subject || obj.Subject || "",
    text: obj.text || "",
    html: obj.html || "",
    date: obj.date || obj.Date || "",
    attachments: [],
  };
}

export async function generateEmail(): Promise<InternalEmailInfo> {
  const login = await fetchJSON<LoginResponse>(`${BASE_URL}/gtoken`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: "{}",
  });
  if (login.status < 200 || login.status >= 300) {
    throw new Error(`ockito gtoken http ${login.status}`);
  }
  const accessToken = String(login.json?.access_token || "").trim();
  const refreshToken = String(login.json?.refresh_token || "").trim();
  if (!accessToken || !refreshToken) {
    throw new Error("ockito: invalid token response");
  }

  const emailResponse = await fetchJSON<EmailResponse>(`${BASE_URL}/email`, {
    headers: {
      Authorization: `Bearer ${accessToken}`,
    },
  });
  if (emailResponse.status < 200 || emailResponse.status >= 300) {
    throw new Error(`ockito email http ${emailResponse.status}`);
  }
  const email = normalizeEmailAddress(emailResponse.json?.email);
  if (!email || !email.includes("@")) {
    throw new Error("ockito: invalid email response");
  }

  return {
    channel: CHANNEL,
    email,
    token: encodeToken(accessToken, refreshToken),
  };
}

export async function getEmails(
  token: string,
  email: string,
): Promise<Email[]> {
  const packed = decodeToken(token);
  const accessTokenState = { accessToken: packed.access_token };
  const address = String(email || "").trim();
  if (!address) {
    throw new Error("ockito: empty email");
  }

  const inboxResponse = await fetchBearerJSON<InboxResponse>(
    `${BASE_URL}/retrieve/${encodeURIComponent(address)}/emails`,
    accessTokenState,
    packed.refresh_token,
  );
  const rows = Array.isArray(inboxResponse?.inbox) ? inboxResponse.inbox : [];
  const out: Email[] = [];

  for (const row of rows) {
    const uid = String(row?.uid || row?.timestamp || "").trim();
    if (!uid) {
      out.push(
        normalizeEmail(
          flattenInboxRow(row as Record<string, unknown>, address),
          address,
        ),
      );
      continue;
    }

    try {
      const detail = await fetchBearerJSON<MailMessageResponse>(
        `${BASE_URL}/retrieve/${encodeURIComponent(address)}/${encodeURIComponent(uid)}`,
        accessTokenState,
        packed.refresh_token,
      );
      out.push(normalizeEmail(flattenMessage(detail, address), address));
    } catch {
      out.push(
        normalizeEmail(
          flattenInboxRow(row as Record<string, unknown>, address),
          address,
        ),
      );
    }
  }

  return out;
}
